### PR TITLE
Tweaks: Add support for updating tweak collections.

### PR DESCRIPTION
**Pull request by barakwei**
@mkimi-lightricks
1. The style is weird, and I adhered to the style in this repo (use of ivars, long lines etc.)
2. I mixed mutable and immutable stuff there IMO, but I didn't want to change too much.
3. KVO files are from https://github.com/facebook/kvocontroller, didn't add submodule.

This is not the last PR, other will probably follow as new features will be needed.

@StatusReport general design review will be nice

### CODE REVIEW

**Comment by barakwei** [comment]
@StatusReport please enable travis CI or circle and I'll move to circle.

File: FBTweak/_FBTweakCollectionViewController.h [hunk 1]
```diff
@@ -15,7 +15,7 @@
 /**
   @abstract Displays configuration options for tweak collections.
  */
-@interface _FBTweakCollectionViewController : UIViewController
+@interface _FBTweakCollectionViewController : UITableViewController
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.h:18]: I suggest to format the files you touch according to our style guide. Cause otherwise it's partially formatted and looks weird. (The reformatting should be done in a separate commit).

File: FBTweak/FBTweakCategory.h [hunk 2]
```diff
@@ -11,30 +11,49 @@
...
+ @abstract Resets all tweaks in \c tweakCollection to their default value.
  */
-@property (nonatomic, copy, readonly) NSString *name;
+- (void)reset;
+
+@optional
+
+/**
+ @abstract Updates \c tweakCollections to the latest value, and calls \c completion when done.
+ If there was an error while updaing, \c error contains the error, and if the update was successful,
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: Usually it's kind of standard way to document this behaviour. `Upon error \c error is set, otherwise \c error is \c nil.`

File: FBTweak/FBTweakCategory.h [hunk 3]
```diff
@@ -11,30 +11,49 @@
...
+ \c error is \c nil.
+ @param completion Completion block to be called when the update is complete.
+ */
+- (void)updateWithCompletion:(void (^)(NSError *error))completion;
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: `error` must be `_Nullable`.

Comment (mkimi-lightricks) [reply FBTweak/FBTweakCategory.h]: @StatusReport usually we prefer to `typedef` block types. Is it a guideline?

File: FBTweak/FBTweakCategory.h [hunk 4]
```diff
@@ -11,30 +11,49 @@
...
+
+/**
+ @abstract A named grouping of collections.
+ */
+@interface FBNativeTweakCategory : NSObject <FBTweakCategory, NSCoding>
 
 /**
-  @abstract The collections contained in this category.
+ @abstract Creates a tweak category.
+ @discussion This is the designated initializer.
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: NS_DESIGNATED_INITIALIZER doesn't work?

File: FBTweak/FBTweakCategory.h [hunk 5]
```diff
@@ -11,30 +11,49 @@
 
 @class FBTweakCollection;
 
+@protocol FBTweakCategory <NSObject>
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: Please add a description.

File: FBTweak/FBTweakCategory.h [hunk 6]
```diff
@@ -11,30 +11,49 @@
...
-@property (nonatomic, copy, readonly) NSString *name;
+- (void)reset;
+
+@optional
+
+/**
+ @abstract Updates \c tweakCollections to the latest value, and calls \c completion when done.
+ If there was an error while updaing, \c error contains the error, and if the update was successful,
+ \c error is \c nil.
+ @param completion Completion block to be called when the update is complete.
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: Please specify whether the \c completion block is executed synchronously or not.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 7]
```diff
@@ -28,11 +30,22 @@ - (instancetype)initWithTweakCategory:(id<FBTweakCategory>)category
...
-    [self.refreshControl addTarget:self
-                            action:@selector(_reloadData)
-                  forControlEvents:UIControlEventValueChanged];
-    [self _reloadData];
+
+    _kvoController = [FBKVOController controllerWithObserver:self];
+    [_kvoController observe:_tweakCategory keyPath:@"tweakCollections"
+                        options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
+                          block:^(id observer, id object, NSDictionary *change) {
+                            dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: indentation.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 8]
```diff
@@ -28,11 +30,22 @@ - (instancetype)initWithTweakCategory:(id<FBTweakCategory>)category
...
-    self.refreshControl = [[UIRefreshControl alloc] init];
-    [self.refreshControl addTarget:self
-                            action:@selector(_reloadData)
-                  forControlEvents:UIControlEventValueChanged];
-    [self _reloadData];
+
+    _kvoController = [FBKVOController controllerWithObserver:self];
+    [_kvoController observe:_tweakCategory keyPath:@"tweakCollections"
+                        options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
+                          block:^(id observer, id object, NSDictionary *change) {
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: Why the compiler doesn't complain that `observer`, `object` and `change` aren't used?

File: FBTweak/_FBTweakCollectionViewController.m [hunk 9]
```diff
@@ -28,11 +30,22 @@ - (instancetype)initWithTweakCategory:(id<FBTweakCategory>)category
...
+                            dispatch_async(dispatch_get_main_queue(), ^{
+                              [self _reloadData];
+                            });
+    }];
+
+    if ([self _supportsUpdates]) {
+      self.refreshControl = [[UIRefreshControl alloc] init];
+      [self.refreshControl addTarget:self
+                              action:@selector(_updateCategory)
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: `action:...` can be moved to the preceding line.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 10]
```diff
@@ -64,6 +77,24 @@ - (void)viewWillDisappear:(BOOL)animated
...
+
+- (void)_updateCategory {
+  self.refreshControl.attributedTitle =
+      [[NSAttributedString alloc] initWithString:@"Refreshing..."];
+  [self.refreshControl beginRefreshing];
+  [self.tweakCategory updateWithCompletion:^(NSError * _Nullable error) {
+      dispatch_async(dispatch_get_main_queue(), ^{
+        [self.refreshControl endRefreshing];
+        if (error) {
+
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: Do what? :)

File: FBTweak/_FBTweakCollectionViewController.m [hunk 11]
```diff
@@ -64,6 +77,24 @@ - (void)viewWillDisappear:(BOOL)animated
...
 }
 
+- (BOOL)_supportsUpdates {
+  return [self.tweakCategory respondsToSelector:@selector(updateWithCompletion:)];
+}
+
+- (void)_updateCategory {
+  self.refreshControl.attributedTitle =
+      [[NSAttributedString alloc] initWithString:@"Refreshing..."];
+  [self.refreshControl beginRefreshing];
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m:91]: Do we want to call it on main thread as well?

**Review by mkimi-lightricks — COMMENTED** [review]
Partial round.

File: FBTweak/FBTweakCategory.h (hunk 3 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: This project doesn't have nullability annotations at all and therefore this is automatically nullable. When I add `_Nullable` in a header file, the compiler get crazy and starts complaining about nullability problems everywhere. In the next PRs I'll make these files more LT-compliant and add nullability annotations.

File: FBTweak/FBTweakCategory.h (hunk 4 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: RIght, this was here before I believe.

File: FBTweak/_FBTweakCollectionViewController.h (hunk 1 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.h:18]: This PR should not include any changes to the style, I agree.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 8 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: I'm guessing this isn't enabled in this project.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 7 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: This is Xcode indentation.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 9 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: done

File: FBTweak/_FBTweakCollectionViewController.m (hunk 11 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m:91]: This method is called on the main thread - it's the observer for the refresh control.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 10 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: I actually didn't know what to do here. Decided to just show an error.

File: FBTweak/FBTweakCategory.h (hunk 6 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: Changed the text

File: FBTweak/FBTweakCategory.h (hunk 2 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: Changed the text

File: FBTweak/FBTweakCategory.h (hunk 5 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: Added

**Comment by barakwei** [comment]
Round complete

**Comment by StatusReport** [comment]
@barakwei configured Circle.

**Comment by StatusReport** [comment]
@barakwei if someone will use `FBKVOController` somewhere else, you'll have a collision, so think if that's what you really want. In any case if you choose to proceed make sure that the `LICENSE` includes `FBKVOController`'s license as well.

File: FBTweak/_FBTweakCollectionViewController.h (hunk 1 above)
Comment (StatusReport) [reply FBTweak/_FBTweakCollectionViewController.h:18]: Changing the style will make it harder to pull changes from upstream. I'm against it unless we explicitly choose to fork this and not pull future changes.

Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.h:18]: For now, we'll keep the existing style. As we continue to hack and add features, we'll consider changing the style, as the origin project seems dead for now, and add we continue to change code, merging from upstream will be more and more difficult.

We can say that new files will be added using LT style.

**Comment by barakwei** [comment]
I tent to prefer adding it as a submodule over the current solution, but... while going over the code, I saw they use KVO [here](https://github.com/Lightricks/Tweaks/blob/master/FBTweak/_FBTweakColorViewController.m#L48), and they just remove the observation in `dealloc`. The KVO in this PR is simple and what they did should suffice. Since what they do seems to work pretty well, I can do the same thing. Thoughts?

File: FBTweak/FBTweakCategory.h [hunk 12]
```diff
@@ -12,29 +12,52 @@
...
 /**
-  @abstract Fetches a collection by name.
-  @param name The collection name to find.
+ @abstract Creates a tweak category.
+ */
+- (instancetype)initWithName:(NSString *)name NS_DESIGNATED_INITIALIZER;
+
+/**
+ @abstract Fetches a collection by name.
+ @param name The collection name to find.
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: * `find` -> `fetch`. 
* Please document what happens if the `name` doesn't exist.

File: FBTweak/FBTweakCategory.m [hunk 13]
```diff
@@ -7,24 +7,27 @@
...
 #import "FBTweakCategory.h"
 #import "FBTweakCollection.h"
 
-@implementation FBTweakCategory {
-  NSMutableArray *_orderedCollections;
-  NSMutableDictionary *_namedCollections;
-}
+@interface FBNativeTweakCategory ()
+
+@property (nonatomic, copy, readwrite) NSMutableArray<FBTweakCollection *> *tweakCollections;
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.m]: There are two options as I see:
* Comment the `tweakCollections`
* Remove the empty lines before and after this line.

I prefer the 1st one.

File: FBTweak/FBTweakCategory.m [hunk 14]
```diff
@@ -34,40 +37,49 @@ - (instancetype)initWithName:(NSString *)name
...
 - (FBTweakCollection *)tweakCollectionWithName:(NSString *)name
 {
-  return _namedCollections[name];
-}
-
-- (NSArray *)tweakCollections
-{
-  return [_orderedCollections copy];
+  NSUInteger index = [_tweakCollections indexOfObjectPassingTest:
+                      ^BOOL(FBTweakCollection * _Nonnull collection, NSUInteger __unused idx, BOOL * _Nonnull stop) {
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.m:58]: `^BOOL(FBTweakCollection * _Nonnull collection, NSUInteger __unused idx, BOOL * _Nonnull stop) {`
->
`^BOOL(FBTweakCollection *collection, NSUInteger, BOOL *) {`

Comment (mkimi-lightricks) [reply FBTweak/FBTweakCategory.m:58]: So what we decided about the indentation of changes you do? Than you use the default Xcode's indentation?

File: FBTweak/FBTweakCategory.m [hunk 15]
```diff
@@ -34,40 +37,49 @@ - (instancetype)initWithName:(NSString *)name
...
+  return index != NSNotFound ? [self.tweakCollections objectAtIndex:index] : nil;
 }
 
 - (void)addTweakCollection:(FBTweakCollection *)tweakCollection
 {
-  [_orderedCollections addObject:tweakCollection];
-  [_namedCollections setObject:tweakCollection forKey:tweakCollection.name];
+  NSMutableArray<FBTweakCollection *> *collections = self.tweakCollections;
+  [collections addObject:tweakCollection];
+  self.tweakCollections = collections;
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.m]: * Why it's required? AFAIU `collections` and `self.teakCollections` are pointer to the same instance of `NSMutableArray`...
* Can following implementation of `addTweakCollection:` fit here?
  ```objc
  [self.tweakCollections addObject:tweakCollection];
  ```

File: FBTweak/FBTweakCategory.m [hunk 16]
```diff
@@ -34,40 +37,49 @@ - (instancetype)initWithName:(NSString *)name
...
 }
 
 - (void)removeTweakCollection:(FBTweakCollection *)tweakCollection
 {
-  [_orderedCollections removeObject:tweakCollection];
-  [_namedCollections removeObjectForKey:tweakCollection.name];
+  NSMutableArray<FBTweakCollection *> *collections = self.tweakCollections;
+  [collections removeObject:tweakCollection];
+  self.tweakCollections = collections;
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.m:74]: Same as above...

File: FBTweak/FBTweakInline.m [hunk 17]
```diff
@@ -137,6 +139,12 @@ + (void)load
       }
     }
   }
+
+  FBTweakStore *store = [FBTweakStore sharedInstance];
+  [categories enumerateKeysAndObjectsUsingBlock:
+   ^(NSString * _Nonnull __unused name, FBNativeTweakCategory * _Nonnull category, BOOL * _Nonnull stop) {
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakInline.m:145]: `^(NSString * _Nonnull __unused name, FBNativeTweakCategory * _Nonnull category, BOOL * _Nonnull stop) {`
->
`^(NSString *, FBNativeTweakCategory *category, BOOL *) {`

File: FBTweak/FBTweakCategory.h [hunk 18]
```diff
@@ -12,29 +12,52 @@
...
+- (void)updateWithCompletion:(void (^)(NSError *error))completion;
+
+@end
 
 /**
-  @abstract The collections contained in this category.
+ @abstract A named grouping of collections.
  */
-@property (nonatomic, copy, readonly) NSArray *tweakCollections;
+@interface FBNativeTweakCategory : NSObject <FBTweakCategory, NSCoding>
```
Comment (mkimi-lightricks) [inline FBTweak/FBTweakCategory.h]: I think there are better names that matches the description `A named grouping of collections` than `FBNativeTweakCategory`.
Perhaps `FBGroupedTeakCategory` or something similar.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 19]
```diff
@@ -7,31 +7,45 @@
...
   if ((self = [super init])) {
     _tweakCategory = category;
+    self.clearsSelectionOnViewWillAppear = YES;
     self.title = _tweakCategory.name;
-    [self _reloadData];
+
+    _kvoController = [FBKVOController controllerWithObserver:self];
+    [_kvoController observe:_tweakCategory keyPath:@"tweakCollections"
+                        options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
+                          block:^(id observer __unused, id object  __unused,
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: remove the non necessary `__unused` along with argument names.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 20]
```diff
@@ -7,31 +7,45 @@
...
+- (instancetype)initWithTweakCategory:(id<FBTweakCategory>)category
 {
   if ((self = [super init])) {
     _tweakCategory = category;
+    self.clearsSelectionOnViewWillAppear = YES;
     self.title = _tweakCategory.name;
-    [self _reloadData];
+
+    _kvoController = [FBKVOController controllerWithObserver:self];
+    [_kvoController observe:_tweakCategory keyPath:@"tweakCollections"
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: Do you have to remove this observation when done?

File: FBTweak/_FBTweakCollectionViewController.m [hunk 21]
```diff
@@ -41,28 +55,17 @@ - (void)viewDidLoad
...
   [super viewDidLoad];
 
-  _tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
-  _tableView.delegate = self;
-  _tableView.dataSource = self;
-  _tableView.autoresizingMask = (UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight);
-  [self.view addSubview:_tableView];
-  
-  self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(_done)];
+  self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: Please undo the formatting.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 22]
```diff
@@ -74,12 +77,48 @@ - (void)viewWillDisappear:(BOOL)animated
...
+        [self.refreshControl endRefreshing];
+        if (error) {
+          [self presentViewController:[self errorAlertControllerWithError:error] animated:YES
+                           completion:nil];
+        }
+      });
+    }];
+}
+
+- (UIAlertController *)errorAlertControllerWithError:(NSError * _Nonnull)error {
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m:103]: Remove the `_Nonnull`.

File: FBTweak/_FBTweakCollectionViewController.m [hunk 23]
```diff
@@ -74,12 +77,48 @@ - (void)viewWillDisappear:(BOOL)animated
...
+  NSString *alertMessage = [NSString stringWithFormat:@"An error occured while updating: %@.",
+                            error.description];
+  UIAlertController *alertController = [UIAlertController
+                                        alertControllerWithTitle:@"Error updating catergory"
+                                        message:alertMessage
+                                        preferredStyle:UIAlertControllerStyleAlert];
+  UIAlertAction *okAction = [UIAlertAction
+                             actionWithTitle:NSLocalizedString(@"OK", @"OK action")
+                             style:UIAlertActionStyleDefault
+                             handler:^(UIAlertAction __unused *action){}];
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m:113]: `UIAlertAction __unused *action` -> `UIAlertAction *`

File: FBTweak/_FBTweakCollectionViewController.m [hunk 24]
```diff
@@ -74,12 +77,48 @@ - (void)viewWillDisappear:(BOOL)animated
...
+
+  [alertController addAction:okAction];
+  return alertController;
+}
+
 - (void)_reloadData
 {
-  _sortedCollections = [_tweakCategory.tweakCollections sortedArrayUsingComparator:^(FBTweakCollection *a, FBTweakCollection *b) {
+  _sortedCollections = [_tweakCategory.tweakCollections sortedArrayUsingComparator:
+                        ^(FBTweakCollection *a, FBTweakCollection *b) {
```
Comment (mkimi-lightricks) [inline FBTweak/_FBTweakCollectionViewController.m]: Undo the formatting.

**Review by mkimi-lightricks — COMMENTED** [review]
Round.

File: FBTweak/FBTweakCategory.h (hunk 12 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.h]: This is copied from the existing class

File: FBTweak/FBTweakCategory.m (hunk 14 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.m:58]: I can't do that, it's cpp, and this file is `.m`.

File: FBTweak/FBTweakCategory.m (hunk 15 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.m]: The problem here is that we now want the categories to be reactive. So the `tweakCollections` property should be KVO-compliant. Your suggestion does call the setter for `tweakCollections` and KVO does not fire.
There are two ways to make this work here:
1. Keep this implementations (or very similar) like
```
  [self.tweakCollections addObject:tweakCollection];
  self.tweakCollections = self.tweakCollections;
```

2. Not use mutable array and then just do
```
self.tweakCollections = [self.tweakCollections arrayByAddingObject:tweakCollection];
```

File: FBTweak/FBTweakCategory.m (hunk 16 above)
Comment (barakwei) [reply FBTweak/FBTweakCategory.m:74]: Same

File: FBTweak/FBTweakInline.m (hunk 17 above)
Comment (barakwei) [reply FBTweak/FBTweakInline.m:145]: Same

File: FBTweak/_FBTweakCollectionViewController.m (hunk 20 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: No, this is done automatically when the controller deallocates (always before this class deallocates)

File: FBTweak/_FBTweakCollectionViewController.m (hunk 19 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: can't do - the same issue about cpp.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 21 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: Done

File: FBTweak/_FBTweakCollectionViewController.m (hunk 22 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m:103]: mmmm in `.m` files apparently I can use nullability annotations, I think we should keep them.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 23 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m:113]: same, can't.

File: FBTweak/_FBTweakCollectionViewController.m (hunk 24 above)
Comment (barakwei) [reply FBTweak/_FBTweakCollectionViewController.m]: Done

**Comment by barakwei** [comment]
Not done with the round, still fiddling with KVO possibilities.

**Comment by barakwei** [comment]
Round. I'm not sure about `FBNativeTweakCollection`. Maybe this should be `FBMutableTweakCollection`.
Let's discuss this F2F, as I have plans to do a similar thing on the `FBTweak` object.

File: FBTweak/FBTweakCategory.m (hunk 15 above)
Comment (mkimi-lightricks) [reply FBTweak/FBTweakCategory.m]: Choose any of the two, and add a comment about it.

Comment (mkimi-lightricks) [reply FBTweak/FBTweakCategory.m]: Same below.

File: FBTweak/FBTweakCategory.h (hunk 12 above)
Comment (mkimi-lightricks) [reply FBTweak/FBTweakCategory.h]: Still the documentation needs to updated, even if it was copied from somewhere else.

**Comment by mkimi-lightricks** [comment]
Round.
Please address the last comments and squash.

**Comment by barakwei** [comment]
Closing this one favor of #5 and #6
